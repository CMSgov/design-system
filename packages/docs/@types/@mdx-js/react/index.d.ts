declare module '@mdx-js/react' {
  const MDXProvider: React.ComponentType;
  export { MDXProvider };
}
