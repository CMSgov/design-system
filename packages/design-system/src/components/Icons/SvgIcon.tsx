import type * as React from 'react';
import classNames from 'classnames';
import useId from '../utilities/useId';

export interface BaseSvgIconProps {
  /**
   * Describes the value of the `aria-hidden` attribute on the SVG. Defaulted to true with the assumption that most icons are decorative.
   * If the icon does not have any associated label text, set this to `false` and ensure a `title` is provided for improved accessibility.
   */
  ariaHidden?: boolean;
  /**
   * Additional CSS classes to be added to the svg element
   */
  className?: string;
  /**
   * The elements that make up the SVG
   */
  children: React.ReactNode;
  /**
   * Long-text description of any SVG. Use for complex icons, otherwise `title` prop will suffice.
   */
  description?: string;
  /**
   * A custom `id` attribute for the SVG
   */
  id?: string;
  /**
   * If `true` sets inverse fill color.
   */
  inversed?: boolean;
  /**
   * The descriptive name for the SVG icon
   */
  title: string;
  /**
   * A string describing the viewbox of the SVG.
   *
   * It is recommended that the icon is centered and fill up the default viewport size.
   * See [this blog post](https://webdesign.tutsplus.com/tutorials/svg-viewport-and-viewbox-for-beginners--cms-30844) for further explanation on viewBox and how to use it.
   */
  viewBox?: string;
}

export type SvgIconProps = BaseSvgIconProps &
  Omit<React.SVGProps<SVGSVGElement>, keyof BaseSvgIconProps>;

// a type for react icon components that makes the 'title' prop optional & removes 'children' from type
export type IconCommonProps = Partial<Omit<SvgIconProps, 'children'>>;

export const SvgIcon = ({
  ariaHidden,
  className,
  children,
  description,
  id,
  inversed,
  title,
  viewBox,
  ...otherProps
}: SvgIconProps) => {
  const svgClasses = classNames('ds-c-icon', { 'ds-c-icon--inverse': inversed }, className);

  const rootId = useId('icon--', id);
  const titleId = `${rootId}__title`;
  const descriptionId = `${rootId}__desc`;
  const ariaLabelledBy = description ? `${titleId} ${descriptionId}` : titleId;
  const isSrVisible = !ariaHidden;
  const screenReaderProps = {};
  if (isSrVisible) {
    screenReaderProps['aria-labelledby'] = ariaLabelledBy;
    screenReaderProps['role'] = 'img';
  }

  return (
    <svg
      aria-hidden={ariaHidden}
      className={svgClasses}
      id={id ?? isSrVisible ? rootId : undefined}
      viewBox={viewBox}
      xmlns="http://www.w3.org/2000/svg"
      {...screenReaderProps}
      {...otherProps}
    >
      {isSrVisible && <title id={titleId}>{title}</title>}
      {isSrVisible && description && <desc id={descriptionId}>{description}</desc>}
      {children}
    </svg>
  );
};

SvgIcon.defaultProps = {
  ariaHidden: true,
  inversed: false,
};
