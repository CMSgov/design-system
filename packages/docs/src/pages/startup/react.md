---
title: Using React components
weight: 3
---

The examples below assume you've installed the design system using NPM and have already setup your build system.

<h4 class="ds-h4 ds-u-font-size--base">Default imports</h4>

Individual components can be imported from their individual export file.

```jsx
import Button from '@cmsgov/design-system-core/dist/components/Button/Button';
```

<h4 class="ds-h4 ds-u-font-size--base">Named imports</h4>

Components can also be imported using the shorter syntax below.

<div class="ds-c-alert ds-c-alert--warn ds-u-margin-bottom--2 ds-u-font-size--small">
  <div class="ds-c-alert__body">
    <h3 class="ds-c-alert__heading ds-u-font-size--base">Performance note</h3>
    <p class="ds-c-alert__text">
      This approach may result in a much larger file than you intend. Depending on what module bundler you use, all of the design system's React components may be included in the bundled file even if you didn't specifically import them. This can be avoided by enabling features like [tree shaking in Webpack](https://webpack.js.org/guides/tree-shaking/).
    </p>
  </div>
</div>

```jsx
import { Button } from '@cmsgov/design-system-core';
```

[View an example](https://github.com/CMSgov/design-system/blob/master/examples/react-app/src/scripts/index.js)

<h2 id="examples" class="ds-h2 ds-u-color--primary-darker">Examples</h2>

Additional examples of the design system in use can be viewed on GitHub. These projects demonstrate the various ways you can incorporate the design system into your development process and various use cases.

<a href="https://github.com/CMSgov/design-system/tree/master/examples/" class="ds-c-button">Browse example projects</a>

<h2 id="need-help" class="ds-h2 ds-u-color--primary-darker">Need help or ran into an issue?</h2>

If you're having trouble installing or setting up the design system, or if you think you've found a bug, feel free to [open an issue on GitHub](https://github.com/CMSgov/design-system/issues).
