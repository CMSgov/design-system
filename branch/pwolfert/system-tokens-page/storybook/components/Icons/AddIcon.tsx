import React from 'react';
import { t } from '../i18n';
import SvgIcon, { IconCommonProps } from './SvgIcon';

const defaultProps = {
  className: '',
  viewBox: '3 3 18 18',
};

function AddIcon(props: IconCommonProps): React.ReactElement {
  const iconCssClasses = `ds-c-icon--add ${props.className || ''}`;

  return (
    <SvgIcon title={t('icons.add')} {...defaultProps} {...props} className={iconCssClasses}>
      <path d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z" />
    </SvgIcon>
  );
}

export default AddIcon;
