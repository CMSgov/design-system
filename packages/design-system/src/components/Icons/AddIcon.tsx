import React from 'react';
import SvgIcon, { IconCommonProps } from './SvgIcon';

const defaultProps = {
  className: '',
  title: 'Add',
  viewBox: '3 3 18 18',
};

function AddIcon(props: IconCommonProps): React.ReactElement {
  const iconCssClasses = `ds-c-icon--add ${props.className || ''}`;

  return (
    <SvgIcon {...defaultProps} {...props} className={iconCssClasses}>
      <path d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z" />
    </SvgIcon>
  );
}

export default AddIcon;
