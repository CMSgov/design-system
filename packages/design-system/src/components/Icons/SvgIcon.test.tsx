import React from 'react';
import { render, screen } from '@testing-library/react';
import { SvgIcon } from './SvgIcon';
import { AddIcon } from './AddIcon';

describe('SvgIcon', () => {
  const renderSvgIcon = (overrideProps?) => {
    return render(
      <SvgIcon ariaHidden={false} title="test icon" {...overrideProps}>
        <path />
      </SvgIcon>
    );
  };

  it('passes through additional props', () => {
    renderSvgIcon({ 'data-testid': 'iconTest', id: 'static-id' });
    const iconEl = screen.getByTestId('iconTest');
    expect(iconEl).toMatchSnapshot();
  });

  it('wrapper icon can pass through additional props', () => {
    render(<AddIcon data-testid="addIconTest" />);
    const iconEl = screen.getByTestId('addIconTest');
    expect(iconEl).toMatchSnapshot();
  });

  describe('when ariaHidden is false', () => {
    it('shows accessibility attributes', () => {
      renderSvgIcon({ id: 'static-id' });
      const iconEl = screen.getByRole('img');

      expect(iconEl).toBeDefined();
      expect(iconEl.getAttribute('aria-labelledby')).toEqual('static-id__title');
    });

    it('renders title', () => {
      renderSvgIcon({ id: 'static-id' });
      const iconEl = screen.getByTitle('test icon');
      expect(iconEl).toBeDefined();
    });

    it('updates aria-labelledby if description exists', () => {
      renderSvgIcon({ description: 'i am a description of the svg', id: 'static-id' });
      const iconEl = screen.getByRole('img');

      expect(iconEl.getAttribute('aria-labelledby')).toEqual('static-id__title static-id__desc');
    });
  });

  describe('when ariaHidden is true', () => {
    it('does not have role="img"', () => {
      renderSvgIcon({ ariaHidden: true });
      const iconEl = screen.queryByRole('img');
      expect(iconEl).toBe(null);
    });

    it('does not have an id', () => {
      renderSvgIcon({ 'data-testid': 'testId', ariaHidden: true });
      const iconEl = screen.getByTestId('testId');
      expect(iconEl).not.toHaveAttribute('id');
    });

    it('renders without title or description"', () => {
      renderSvgIcon({ ariaHidden: true, 'data-testid': 'testId' });
      const iconEl = screen.getByTestId('testId');
      expect(iconEl).toMatchSnapshot();
    });
  });
});
