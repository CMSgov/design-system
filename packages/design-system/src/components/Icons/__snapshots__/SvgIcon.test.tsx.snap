// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SvgIcon passes through additional props 1`] = `
<svg
  aria-hidden="false"
  aria-labelledby="static-id__title"
  class="ds-c-icon"
  data-testid="iconTest"
  id="static-id"
  role="img"
  xmlns="http://www.w3.org/2000/svg"
>
  <title
    id="static-id__title"
  >
    test icon
  </title>
  <path />
</svg>
`;

exports[`SvgIcon when ariaHidden is true renders without title or description" 1`] = `
<svg
  aria-hidden="true"
  class="ds-c-icon"
  data-testid="testId"
  xmlns="http://www.w3.org/2000/svg"
>
  <path />
</svg>
`;

exports[`SvgIcon wrapper icon can pass through additional props 1`] = `
<svg
  aria-hidden="true"
  class="ds-c-icon ds-c-icon--add "
  data-testid="addIconTest"
  viewBox="3 3 18 18"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
  />
</svg>
`;
